numpy.h indirectly defines `I` - breaks following includes

The header `pybind11/numpy.h` includes `complex.h` which, at least in glibc, defines `I` as a preprocessor macro to denote the imaginary unit. Unfortunately `I` is a popular name for template parameters in e.g. the boost library. Therefore, the following two lines will fail to compile

```
#include <pybind11/numpy.h>
#include <boost/variant.hpp>
```

while the following three lines will work fine.

```
#include <pybind11/numpy.h>
#undef I
#include <boost/variant.hpp>
```

The compiler error is very long, however it starts with

```
include/boost/move/iterator.hpp:151:11: error: declaration of anonymous class must be a definition
template <class I>
```

which indicates that the `I` is the problem.

It took me some time to find the cause for that particular error. Do you think it would be reasonable to `#undef I` within `pybind11/numpy.h` to spare others from the same problem?

--- a/include/pybind11/complex.h
+++ b/include/pybind11/complex.h
@@ -12,6 +12,11 @@
 #include "pybind11.h"
 #include <complex>
 
+/// glibc defines I as a macro which breaks things, e.g., boost template names
+#ifdef I
+#  undef I
+#endif
+
 NAMESPACE_BEGIN(pybind11)
 
 PYBIND11_DECL_FMT(std::complex<float>, "Zf");

